src/Services/Virtualization/PortalAuthenticationModule.cs

```diff
@@ -20,158 +21,80 @@
-                                // Get the login page Url (eg. http://localhost:1315/home/login)
-                                string loginPageUrl = currentPortalContext.GetLoginPageUrl();
-                                // Append trailing slash
-                                if (loginPageUrl != null && !loginPageUrl.EndsWith("/"))
-                                    loginPageUrl = loginPageUrl + "/";
-
-                                // Cut down the querystring (eg. drop ?Param1=value1@Param2=value2)
-                                string currentRequestUrlWithoutQueryString = currentPortalContext.RequestedUri.GetComponents(UriComponents.Scheme | UriComponents.Host | UriComponents.Port | UriComponents.Path, UriFormat.Unescaped);
-
-                                // Append trailing slash
-                                if (!currentRequestUrlWithoutQueryString.EndsWith("/"))
-                                    currentRequestUrlWithoutQueryString = currentRequestUrlWithoutQueryString + "/";
-
-                                // Redirect to the login page, if neccessary.
-                                if (currentRequestUrlWithoutQueryString != loginPageUrl)
-                                    context.Response.Redirect(loginPageUrl + "?OriginalUrl=" + System.Web.HttpUtility.UrlEncode(currentPortalContext.RequestedUri.ToString()), true);
-                            }
-                            break;
-                        default:
-                            AuthenticationHelper.DenyAccess(context);
-                            break;
+                        _securityKey = EncryptionHelper.CreateSymmetricKey(Configuration.TokenAuthentication.SymmetricKeySecret);
                     }
+                    return _securityKey;
                 }
             }
         }
 
-        private bool DispatchBasicAuthentication(HttpApplication application, HttpRequest request)
+        public Func<object, HttpContextBase> GetContext = (sender) => new HttpContextWrapper(((HttpApplication) sender).Context);
+        public Func<object, HttpRequestBase> GetRequest = (sender) => new HttpRequestWrapper(((HttpApplication)sender).Context.Request);
+        public Func<object, HttpResponseBase> GetResponse = (sender) => new HttpResponseWrapper(((HttpApplication)sender).Context.Response);
+        public Func<IPrincipal> GetVisitorPrincipal = () => new PortalPrincipal(User.Visitor);
+        public Func<string, IPrincipal> LoadUserPrincipal = (userName) => new PortalPrincipal(User.Load(userName));
+        public Func<string, string, bool> IsUserValid = (userName, password) => Membership.ValidateUser(userName, password);
+        public Func<IDisposable> GetSystemAccount = () => new SystemAccount();
+        public Func<string> GetBasicAuthHeader = () => PortalContext.Current.BasicAuthHeaders;
+
+        private bool _anonymAuthenticated;

```

Comment:
WARNING: this is very risky! Modules are **instantiated only once** during the app life cycle, they are technically singleton instances, **we must not use instance variables**, because the module is re-used across requests. We have to find another way to pass this info through methods.